tests/generation/test_utils.py

```diff
@@ -1236,6 +1236,69 @@ def test_prompt_lookup_decoding_matches_greedy_search(self):
             for output in (output_greedy, output_prompt_lookup):
                 self._check_outputs(output, input_ids, model.config, use_cache=True)
 
+    def test_dola_decoding_sample(self):
+        # TODO (joao): investigate skips, try to reduce incompatibilities
+        for model_class in self.all_generative_model_classes:
+            if any(model_name in model_class.__name__.lower() for model_name in ["reformer"]):
+                self.skipTest("Skip Reformer as the lm_head input size is 2 * hidden size, adopted from Rev Nets.")
+            if any(
+                model_name in model_class.__name__.lower()
+                for model_name in [
+                    "wav2vec",
+                    "clvp",
+                    "bark",
+                ]
+            ):
+                self.skipTest("Skip speech models")
+            if any(
+                model_name in model_class.__name__.lower()
+                for model_name in [
+                    "marian",
+                    "mbart",
+                    "pegasus",
+                ]
+            ):
+                self.skipTest("DoLa is not supported for the models without returning layerwise hidden states")
+
+            # enable cache if the model is not openai-gpt, xlnet, cpm, or xlm
+            config, input_ids, attention_mask, _ = self._get_input_ids_and_config(batch_size=1)
+            config.use_cache = True
+            if any(
+                model_name in model_class.__name__.lower()
+                for model_name in [
+                    "openaigpt",
+                    "xlnet",
+                    "cpm",
+                    "xlm",
+                ]
+            ):
+                config.use_cache = False
```

Comment:
It's because some old models like `xlm` does not keep `past_key_values` at all, so if we set `config.use_cache = True` here, the check will fail when checking the shape of`past_key_values`. Set `config.use_cache = False` for these models can prevent this issue.

To make it not dependent on the model names, I change this part into:

            if not hasattr(config, "use_cache"):
                config.use_cache = False
            else:
                config.use_cache = True

This new version makes more sense. If the model config doesn't have the attribute of`config.use_cache`, they would not maintain `past_key_values` and thus we should just test with `config.use_cache = False`. Otherwise, set it to `True` to test decoding with cache!